File: 26/main.py
######################################################################
###############	26. Remove Duplicates From Array	##############
######################################################################


'''
Replace the duplicate with last unique elements and repeat till EOA
Time Complexity: O(n) n is the length of array and each element is visited once
Space Complexity:O(1) as we only keep constant space pointers in
'''

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        
        prev = 0
        currIndex = 0
        
        for i,val in enumerate(nums):
            if i == 0:
                prev = val
                currIndex = 1
            else:
                if val == prev:
                    continue
                else:
                    nums[currIndex] = val
                    prev = val
                    currIndex += 1
                
                    
        return currIndex

File: 26/main.java
/***
*
*	26. Remove Duplicates from Sorted Array
*
**/

/*
Just replace the elements with non-duplicates
Time Complexity:O(N) number of elements in array is N and visit them once
Space Complexity:O(1) constant space is required to store pointer
*/

class Solution {
public int removeDuplicates(int[] nums) {
    if (nums.length == 0) return 0;
    int i = 0;
    for (int j = 1; j < nums.length; j++) {
        if (nums[j] != nums[i]) {
            i++;
            nums[i] = nums[j];
        }
    }
    return i + 1;
}
}
